### PR TITLE
1557048: Add reason codes to the metrics ping

### PR DESCRIPTION
**Pull request by mdboom**
Requires https://github.com/mozilla/glean_parser/pull/157 and https://github.com/mozilla-services/mozilla-pipeline-schemas/pull/484 to be merged first.

This was a real iceberg.

~Also, no Swift or Python support yet, but thought it worth getting feedback on this much first.~

### CODE REVIEW

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt [hunk 1]
```diff
@@ -187,9 +198,11 @@ internal class MetricsPingScheduler(
         // schedule the metrics ping for immediate collection. We only need to perform
         // this check at startup (when overduePingAsFirst is true).
         if (isDifferentVersion()) {
+            Log.i(LOG_TAG, "The application just updated. Send metrics ping now.")
+
             @Suppress("EXPERIMENTAL_API_USAGE")
             Dispatchers.API.executeTask {
-                collectPingAndReschedule(now, startupPing = true)
+                collectPingAndReschedule(now, startupPing = true, reason = Pings.metricsReasonCodes.upgrade)
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt]: Let's also update [this doc](https://mozilla.github.io/glean/book/user/pings/metrics.html) mentioning the reasons. We should probably also cross-check if the list of edge cases is still correct and potentially add the reasons there, too.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt [hunk 2]
```diff
@@ -256,7 +272,7 @@ internal class MetricsPingScheduler(
      * @param now a [Calendar] instance representing the current time.
      */
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun collectPingAndReschedule(now: Calendar, startupPing: Boolean = false) {
+    internal fun collectPingAndReschedule(now: Calendar, startupPing: Boolean, reason: Pings.metricsReasonCodes) {
         Log.i(LOG_TAG, "Collecting the 'metrics' ping, now = ${now.time}, startup = $startupPing")
```
[the remark this answers is not in the document]
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt]: Good idea

**Comment by mdboom** [comment]
@travis79 : Any thoughts on the failure in the iOS integration test?  I can't reproduce locally, i.e.:

```
cd samples/ios/app
carthage bootstrap --platform iOS --cache-builds --configuration Debug --verbose
```

works for me.  I suspect this has something to do with how the integration test build is "special"?

File: glean-core/ios/Glean/Glean.swift [hunk 3]
```diff
@@ -232,7 +232,7 @@ public class Glean {
     ///
     /// If the ping currently contains no content, it will not be assembled and
     /// queued for sending.
-    func submitPingsByName(pingNames: [String]) {
+    func submitPingsByName(pingNames: [String], reason _: String? = nil) {
```
Comment (travis79) [inline glean-core/ios/Glean/Glean.swift]: Are you sure you meant to have `reason _: String? = nil`?  This means that if someone passes a `reason` to the function, that the function should ignore it and basically doesn't provide access to it inside of the function.

Simply having the default value will allow you to omit it if there is no reason.
```suggestion
    func submitPingsByName(pingNames: [String], reason: String? = nil) {
```

File: glean-core/ios/Glean/Metrics/Ping.swift [hunk 4]
```diff
@@ -4,22 +4,55 @@
...
+///
+/// An enum with no values for convenient use as the default set of extra keys
+/// that an `EventMetricType` can accept.
+public enum NoReasonCodes: ReasonCodes {
+    public func index() -> Int {
+        return 0
+    }
+}
+
+public class PingBase {
```
Comment (travis79) [inline glean-core/ios/Glean/Metrics/Ping.swift]: I think changing this to `PingBase` may be why the integration tests are failing.  It may take a tweak to the Xcode build to update the include to make this work. (Sorry, it's not ideal, I know).  Let me take a deeper look tomorrow and see if I can help.

File: glean-core/ios/Glean/Glean.swift (hunk 3 above)
Comment (mdboom) [reply glean-core/ios/Glean/Glean.swift]: Weird -- the Swift formatter told me to add it (and failed without it).  But I agree, that doesn't make much sense.

File: glean-core/ios/Glean/Metrics/Ping.swift (hunk 4 above)
Comment (mdboom) [reply glean-core/ios/Glean/Metrics/Ping.swift]: Thanks.  No rush on this!

File: glean-core/ios/Glean/Glean.swift (hunk 3 above)
Comment (travis79) [reply glean-core/ios/Glean/Glean.swift]: Since `reason` isn't used in the function, the formatter thought it was being helpful by suggesting you make it explicit that you weren't using it.

Comment (travis79) [reply glean-core/ios/Glean/Glean.swift]: I'm guessing it was supposed to be passed to the other function?

File: glean-core/ios/Glean/Metrics/Ping.swift (hunk 4 above)
Comment (travis79) [reply glean-core/ios/Glean/Metrics/Ping.swift]: I took another look at this and the only thing I can think of is that Carthage is not getting the iOS version of Glean with the changes.  I'm a bit stumped as to why this isn't working and I'm trying to set up a build locally that uses a locally built Glean without Carthage to see if it still fails to build.  @badboy do you have any ideas?

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,7 +8,11 @@
     in the baseline ping ([`glean.baseline.locale`](https://github.com/mozilla/glean/blob/c261205d6e84d2ab39c50003a8ffc3bd2b763768/glean-core/metrics.yaml#L28-L42))
     is redundant and will be removed by the end of the quarter.
   * Drop the Glean handle and move state into glean-core ([#664](https://github.com/mozilla/glean/pull/664))
-  * If an experiment includes no `extra` fields, it will no longer include `{"extra": null}` in the JSON paylod.
+  * If an experiment includes no `extra` fields, it will no longer include `{"extra": null}` in the JSON payload.
+  * Support for ping `reason` codes was added.
+  * The metrics ping will now include `reason` codes that indicate why it was
+    sent.
```
Comment (Dexterp37) [inline CHANGELOG.md]:
```suggestion
    submitted.
```

File: docs/user/pings/index.md [hunk 6]
```diff
@@ -31,6 +31,7 @@ Optional fields are marked accordingly.
 | `experiments` | Object | *Optional*. A dictionary of [active experiments](#the-experiments-object) |
 | `start_time` | Datetime | The time of the start of collection of the data in the ping, in local time and with minute precision, including timezone information. |
 | `end_time` | Datetime | The time of the end of collection of the data in the ping, in local time and with minute precision, including timezone information. This is also the time this ping was generated and is likely well before ping transmission time. |
+| `reason` | String | The optional reason the ping was sent. The specific set of values and their meanings are defined for each metric type in the `reasons` field in the `pings.yaml` file. |
```
Comment (Dexterp37) [inline docs/user/pings/index.md]:
```suggestion
| `reason` | String | The optional reason the ping was submitted. The specific set of values and their meanings are defined for each metric type in the `reasons` field in the `pings.yaml` file. |
```

File: docs/user/pings/metrics.md [hunk 7]
```diff
@@ -1,29 +1,36 @@
...
 
 ## Description
-The `metrics` ping is intended for all of the metrics that are explicitly set by the application or are included in the application's `metrics.yaml` file (except events).
-The reported data is tied to the ping's *measurement window*, which is the time between the collection of two `metrics` ping.
-Ideally, this window is expected to be about 24 hours, given that the collection is scheduled daily at 4AM.
-Data in the [`ping_info`](index.md#the-ping_info-section) section of the ping can be used to infer the length of this window.
+The `metrics` ping is intended for all of the metrics that are explicitly set by the application or are included in the application's `metrics.yaml` file (except events). 
+The reported data is tied to the ping's *measurement window*, which is the time between the collection of two `metrics` ping. 
+Ideally, this window is expected to be about 24 hours, given that the collection is scheduled daily at 4AM. 
+However, the metrics ping is only sent while the application is actually running, so in practice, it may not meet the 4AM target very frequently.
```
Comment (Dexterp37) [inline docs/user/pings/metrics.md]:
```suggestion
However, the metrics ping is only submitted while the application is actually running, so in practice, it may not meet the 4AM target very frequently.
```

File: docs/user/pings/metrics.md [hunk 8]
```diff
@@ -1,29 +1,36 @@
...
+Data in the [`ping_info`](index.md#the-ping_info-section) section of the ping can be used to infer the length of this window and the reason that triggered the ping to be sent.
```
Comment (Dexterp37) [inline docs/user/pings/metrics.md]:
```suggestion
Data in the [`ping_info`](index.md#the-ping_info-section) section of the ping can be used to infer the length of this window and the reason that triggered the ping to be submitted.
```

File: docs/user/pings/metrics.md [hunk 9]
```diff
@@ -1,29 +1,36 @@
...
 If the application crashes, unsent recorded metrics are sent along with the next `metrics` ping.
 
+Additionally, it is undesirable to mix metric recording from different versions of the application. Therefore, if a version upgrade is detected, the `metrics` ping is collected immediately before further metrics from the new version are recorded.
+
 > **Note:** As the `metrics` ping was specifically designed for mobile operating systems, it is not sent when using the Glean Python bindings.
 
 ## Scheduling
-The desired behavior is to collect the ping at the first available opportunity after 4AM local time on a new calendar day.
+The desired behavior is to collect the ping at the first available opportunity after 4AM local time on a new calendar day, but given constraints of the platform, it can only be sent while the application is running.
```
Comment (Dexterp37) [inline docs/user/pings/metrics.md]:
```suggestion
The desired behavior is to collect the ping at the first available opportunity after 4AM local time on a new calendar day, but given constraints of the platform, it can only be submitted while the application is running. 
```

File: docs/user/pings/metrics.md [hunk 10]
```diff
@@ -1,29 +1,36 @@
...
-3. the application was open and the 4AM due time was hit.
+2. the application was just upgraded (the version of the app is different from the last time the app was run);
+3. the application was just started (after a crash or a long inactivity period);
+4. the application was running at 4AM.
+
+In the first case, since the application was just installed, if the due time for the current calendar day has passed, a `metrics` ping is immediately generated and scheduled for sending (reason code `overdue`). Otherwise, if the due time for the current calendar day has not passed, a ping collection is scheduled for that time (reason code `today`). 
 
-In the first case, since the application was just installed, if the due time for the current calendar day has passed, a `metrics` ping is immediately generated and scheduled for sending. Otherwise, if the due time for the current calendar day has not passed, a ping collection is scheduled for that time.
+In the second case, if a version change is detected at startup, the metrics ping is immediately sent so that metrics from one version are not aggregated with metrics from another version (reason code `upgrade`).
```
Comment (Dexterp37) [inline docs/user/pings/metrics.md]:
```suggestion
In the second case, if a version change is detected at startup, the metrics ping is immediately submitted so that metrics from one version are not aggregated with metrics from another version (reason code `upgrade`).
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 11]
```diff
@@ -490,11 +490,12 @@ open class GleanInternalAPI internal constructor () {
      * queued for sending.
      *
      * @param pings List of pings to submit.
+     * @param reason The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]:
```suggestion
     * @param reason The reason the ping is being submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 12]
```diff
@@ -512,11 +513,12 @@ open class GleanInternalAPI internal constructor () {
      * file.
      *
      * @param pingNames List of ping names to submit.
+     * @param reason The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]:
```suggestion
     * @param reason The reason the ping is being submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 13]
```diff
@@ -534,8 +536,9 @@ open class GleanInternalAPI internal constructor () {
...
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]:
```suggestion
     * @param reason The reason the ping is being submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/private/PingType.kt [hunk 14]
```diff
@@ -48,9 +63,12 @@ class PingType(
      * There are no guarantees that this will happen immediately.
      *
      * If the ping currently contains no content, it will not be queued.
+     *
+     * @param reason The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/private/PingType.kt]:
```suggestion
     * @param reason The reason the ping is being submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt [hunk 15]
```diff
@@ -63,9 +70,14 @@ internal class MetricsPingScheduler(
      *        or to attempt to schedule it for the current calendar day. If the latter and
      *        we're overdue for the expected collection time, the task is scheduled for immediate
      *        execution.
+     * @param reason The reason the metrics ping is being sent.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt]:
```suggestion
     * @param reason The reason the metrics ping is being submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt [hunk 16]
```diff
@@ -254,10 +271,16 @@ internal class MetricsPingScheduler(
      * next collection.
      *
      * @param now a [Calendar] instance representing the current time.
+     * @param startupUp When true, this is a startup ping that should be sent synchronously.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt]:
```suggestion
     * @param startupUp When true, this is a startup ping that should be submitted synchronously.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt [hunk 17]
```diff
@@ -254,10 +271,16 @@ internal class MetricsPingScheduler(
...
+     * @param reason The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt]:
```suggestion
     * @param reason The reason the ping is being submitted.
```

File: glean-core/ios/Glean/Glean.swift [hunk 18]
```diff
@@ -307,32 +307,36 @@ public class Glean {
     ///
     /// - parameters:
     ///     * pingNames: List of ping names to send
+    ///     * reason: The reason the ping is being sent. Must be one of the strings
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Glean.swift]:
```suggestion
    ///     * reason: The reason the ping is being submitted. Must be one of the strings
```

File: glean-core/ios/Glean/Glean.swift [hunk 19]
```diff
@@ -307,32 +307,36 @@ public class Glean {
...
-            self.submitPingsByNameSync(pingNames: pingNames)
+            self.submitPingsByNameSync(pingNames: pingNames, reason: reason)
         }
     }
 
     /// Collect and submit a list of pings by name for eventual uploading, synchronously
     ///
     /// - parameters:
     ///     * pingNames: List of ping names to send
+    ///     * reason: The reason the ping is being sent. Must be one of the strings
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Glean.swift]:
```suggestion
    ///     * reason: The reason the ping is being submitted. Must be one of the strings
```

File: glean-core/ios/Glean/Metrics/Ping.swift [hunk 20]
```diff
@@ -38,8 +72,15 @@ public class Ping {
     /// There are no guarantees that this will happen immediately.
     ///
     /// If the ping currently contains no content, it will not be queued.
-    public func submit() {
-        Glean.shared.submitPings([self])
+    ///
+    /// - parameters:
+    ///     * reason: The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Metrics/Ping.swift]:
```suggestion
    ///     * reason: The reason the ping is being submitted.
```

File: glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift [hunk 21]
```diff
@@ -29,7 +29,12 @@ class MetricsPingScheduler {
     ///                            or to attempt to schedule it for the current calendar day. If the latter and
     ///                            we're overdue for the expected collection time, the task is scheduled for
     ///                            immediate execution.
-    func schedulePingCollection(_ now: Date, sendTheNextCalendarDay: Bool) {
+    ///     * reason: The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift]:
```suggestion
    ///     * reason: The reason the ping is being submitted.
```

File: glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift [hunk 22]
```diff
@@ -146,21 +159,35 @@ class MetricsPingScheduler {
...
         }
     }
 
     /// Triggers the collection of the "metrics" ping and schedules the next collection.
     ///
     /// - parameters:
     ///     * now: A `Date` representing the current time
-    func collectPingAndReschedule(_ now: Date, startupPing: Bool = false) {
-        logger.info("Collecting the 'metrics' ping, now = \(now), startup = \(startupPing)")
+    ///     * reason: The reason the ping is being sent.
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift]:
```suggestion
    ///     * reason: The reason the ping is being submitted.
```

File: glean-core/metrics.yaml [hunk 23]
```diff
@@ -208,6 +208,20 @@ glean.internal.metrics:
       - glean-team@mozilla.com
     expires: never
 
+  ping_reason:
+    type: string
+    description: |
+      The optional reason the ping was sent.
```
Comment (Dexterp37) [inline glean-core/metrics.yaml]:
```suggestion
      The optional reason the ping was submitted.
```

File: glean-core/pings.yaml [hunk 24]
```diff
@@ -39,8 +39,23 @@ metrics:
     - https://bugzilla.mozilla.org/1512938
   data_reviews:
     - https://bugzilla.mozilla.org/show_bug.cgi?id=1512938#c3
+    - https://bugzilla.mozilla.org/show_bug.cgi?id=1557048#c13
   notification_emails:
     - glean-team@mozilla.com
+  reasons:
+    overdue: |
+      The ping wasn't sent on the current calendar day, but it's after 4am, so
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      The ping wasn't submitted on the current calendar day, but it's after 4am, so
```

File: glean-core/pings.yaml [hunk 25]
```diff
@@ -39,8 +39,23 @@ metrics:
...
+      it's sent immediately
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      it's submitted immediately
```

File: glean-core/pings.yaml [hunk 26]
```diff
@@ -39,8 +39,23 @@ metrics:
...
+    today: |
+      The ping wasn't sent on the current calendar day, but it is still before
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      The ping wasn't submitted on the current calendar day, but it is still before
```

File: glean-core/pings.yaml [hunk 27]
```diff
@@ -39,8 +39,23 @@ metrics:
...
+      4am, so schedule to send on the current calendar day at 4am.
+    tomorrow: |
+      The ping was already sent on the current calendar day, so schedule for
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      The ping was already submitted on the current calendar day, so schedule for
```

File: glean-core/pings.yaml [hunk 28]
```diff
@@ -39,8 +39,23 @@ metrics:
...
+      the next calendar day at 4am.
+    upgrade: |
+      The ping was sent at startup because the application was just upgraded.
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      The ping was submitted at startup because the application was just upgraded.
```

File: glean-core/pings.yaml [hunk 29]
```diff
@@ -39,8 +39,23 @@ metrics:
...
+    reschedule: |
+      The ping was just sent. Rescheduled for the next calendar day at 4am.
```
Comment (Dexterp37) [inline glean-core/pings.yaml]:
```suggestion
      The ping was just submitted. Rescheduled for the next calendar day at 4am.
```

File: glean-core/python/glean/glean.py [hunk 30]
```diff
@@ -381,14 +387,15 @@ def _submit_pings(cls, pings: List["PingType"]):
 
         Args:
             pings (list of PingType): List of pings to submit.
+            reason (str, optional): The reason the ping was sent.
```
Comment (Dexterp37) [inline glean-core/python/glean/glean.py]:
```suggestion
            reason (str, optional): The reason the ping was submitted.
```

File: glean-core/python/glean/metrics/ping.py [hunk 31]
```diff
@@ -32,10 +46,18 @@ def name(self):
...
 
-    def submit(self):
+    def submit(self, reason: Optional[int] = None):
         """
         Collect and submit the ping for eventual uploading.
 
         If the ping currently contains no content, it will not be sent.
+
+        Args:
+            reason (enum, optional): The reason the ping was sent.
```
Comment (Dexterp37) [inline glean-core/python/glean/metrics/ping.py]:
```suggestion
            reason (enum, optional): The reason the ping was submitted.
```

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 32]
```diff
@@ -534,8 +536,9 @@ open class GleanInternalAPI internal constructor () {
...
      */
-    internal fun submitPingsByNameSync(pingNames: List<String>) {
+    internal fun submitPingsByNameSync(pingNames: List<String>, reason: String? = null) {
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Please note that `pingNames` can potentially contain names of different pings (e.g. `baseline`, `events`) and we're only passing a single reason: we should probably have a list of pairs or a map, or two lists in order to match specific optional reasons to specific ping names.

File: glean-core/ios/Glean/Metrics/Ping.swift [hunk 33]
```diff
@@ -4,22 +4,56 @@
 
 import Foundation
 
+public protocol ReasonCodes: Hashable {
+    /// The index of the extra key, used to index the string array passed at `EventMetricType` instantiation.
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/Ping.swift]: This comment is inaccurate.

File: glean-core/ios/Glean/Metrics/Ping.swift [hunk 34]
```diff
@@ -4,22 +4,56 @@
...
+    func index() -> Int
+}
+
+/// Default of no extra keys for events.
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/Ping.swift]: This comment block is inaccurate.

File: glean-core/ios/Glean/Metrics/Ping.swift (hunk 4 above)
Comment (badboy) [reply glean-core/ios/Glean/Metrics/Ping.swift]: We might just be generating invalid Swift code with glean_parser.
Last time I looked at it it wasn't yet fully updated to accomodate the reason codes.

I would need to look into the generated code to determine what's happening.

Comment (mdboom) [reply glean-core/ios/Glean/Metrics/Ping.swift]: glean_parser should be generating valid code now -- it works in the tests of Glean iOS proper, and we are already using this feature there.

Comment (badboy) [reply glean-core/ios/Glean/Metrics/Ping.swift]: In that case I need to take a closer look

File: CHANGELOG.md [hunk 35]
```diff
@@ -8,7 +8,11 @@
...
+  * The version of `glean_parser` has been upgraded to 1.17.0
```
[the remark this answers is not in the document]
Comment (mdboom) [reply CHANGELOG.md]: That should already be done -- the required swift changes to glean_parser have already landed in glean_parser 1.17.0.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/private/PingType.kt [hunk 36]
```diff
@@ -4,29 +4,44 @@
...
 package mozilla.telemetry.glean.private
 
+import com.sun.jna.StringArray
 import mozilla.telemetry.glean.Glean
 import mozilla.telemetry.glean.rust.LibGleanFFI
 import mozilla.telemetry.glean.rust.toByte
 
+/**
+ * The base class of all PingTypes with just enough to track their registration
+ * (and work around the limitations of Kotlin generics).
```
[the remark this answers is not in the document]
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/private/PingType.kt]: Let me reword this.  It's not really an unusual limitation.  We need to create a heterogenous list of `PingType<T>`.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt [hunk 37]
```diff
@@ -423,10 +423,11 @@ class GleanTest {
         ))
 
         val pingName = "custom_ping_1"
-        val ping = PingType(
+        val ping = PingType<NoExtraKeys>(
```
[the remark this answers is not in the document]
Comment (mdboom) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt]: Sure -- a new empty enum just for reason codes probably makes sense.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 32 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: This is a good point, and I agree this is a problem.  Passing in "parallel lists", one of which is optional, seems kind of a strange API though (at least it's internal API).

I actually feel like it might be better to remove all of the plural submit "pings" in favor of a single submit "ping" on all of these functions.  We only use this to submit multiple pings in one place, and we don't even get any parallelism out of thing by batching them up like this.  I think that would be a useful simplification.

Thoughts?  @Dexterp37, @badboy, @travis79, @brizental ?

Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: I guess, specifically, what I'd propose is to merge this PR as-is with respect to this, and as follow-on work, replace `submitPings*` with `submitPing`, which would fix the ambiguity around reason codes.

Comment (badboy) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: One other reason we went with `submitPings` (multiple) was to only call into Rust once.
So far no one has ever complained about FFI calls and given how infrequent and with how small of a list we call it I'm fine with dropping that API

(Users can only submit one ping at a time anyway, through the `CustomPing.submit()` function, right?)

Comment (travis79) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: I'm all for the one at a time, one ping with one reason passed into the function.  I also like the idea of this simplifying the underlying FFI function.

Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: @badboy: Yes, though the public API, users can only submit one ping at once.  The only place where we submit multiple is on going to background (to send baseline and events).

Comment (Dexterp37) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: I'm fine with dropping the plural, but I think it needs to be part of this PR, since this makes #687 buggy (unless we opt for sending the ping on background without a "reason" code.) :-)

Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Ah, good point about #687.  I was mainly just trying to not make this PR any bigger than it already is.  But it should be straightforward.  I'll take it on.

**Comment by mdboom** [comment]
I think this beast is finally ready for a final review.  Apologies for how out-of-hand it became.

File: glean-core/ios/Glean/Metrics/Ping.swift (hunk 4 above)
Comment (badboy) [reply glean-core/ios/Glean/Metrics/Ping.swift]: It seems to be fixed.

File: docs/user/pings/custom.md [hunk 38]
```diff
@@ -155,7 +158,9 @@ Pings.search.submit()
 ```swift
 import Glean
 
-GleanMetrics.Pings.shared.search.submit()
+GleanMetrics.Pings.shared.search.submit(
+    reason: GleanMetrics.Pings.shared.SearchReasonCodes.performed
```
Comment (badboy) [inline docs/user/pings/custom.md]: It's an enum, which means Swift has a nice short form of this:

```swift
GleanMetrics.Pings.shared.search.submit(reason: .performed)
```

(didn't test it, but we do the same on events and their extra keys)

Comment (mdboom) [reply docs/user/pings/custom.md]: Cool.  I can update.

**Comment by mdboom** [comment]
> LGTM. What was that bad failure about? What was the fix?

I was never able to see the segfault again.  The other intermittent was unrelated to this PR, it turned out, and fixed in #692.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt [hunk 39]
```diff
@@ -397,15 +397,15 @@ class LabeledMetricTypeTest {
 
     @Test(expected = IllegalStateException::class)
     fun `Test that labeled events are an exception`() {
-        val eventMetric = EventMetricType<NoExtraKeys>(
+        val eventMetric = EventMetricType<NoReasonCodes>(
```
Comment (travis79) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt]: Did you mean to change this here?  I think `NoExtraKeys` is correct for an `EventMetricType`.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt [hunk 40]
```diff
@@ -397,15 +397,15 @@ class LabeledMetricTypeTest {
...
             disabled = false,
             category = "telemetry",
             lifetime = Lifetime.Application,
             name = "labeled_event_metric",
             sendInPings = listOf("metrics")
         )
 
-        val labeledEventMetric = LabeledMetricType<EventMetricType<NoExtraKeys>>(
+        val labeledEventMetric = LabeledMetricType<EventMetricType<NoReasonCodes>>(
```
Comment (travis79) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt]: I don't think you meant to change this here, I think `NoExtraKeys` is correct for an `EventMetricType`.

**Comment by mdboom** [comment]
Good catch. :+1: on fixing that.

File: glean-core/pings.yaml [hunk 41]
```diff
@@ -39,8 +39,24 @@ metrics:
...
+    - https://bugzilla.mozilla.org/show_bug.cgi?id=1557048#c13
   notification_emails:
     - glean-team@mozilla.com
+  reasons:
+    overdue: |
+      The ping wasn't submitted on the current calendar day, but it's after 4am,
+      so it's submitted immediately
+    today: |
+      The ping wasn't submitted on the current calendar day, but it is still
+      before 4am, so schedule to send on the current calendar day at 4am.
```
Comment (chutten) [inline glean-core/pings.yaml]: The tenses of these are a little weird. Shouldn't they be describing the ping they're on, not the scheduling scenario? "The ping wasn't submitted..." but then how did we get it? The ping must have been submitted.

Maybe reword a la: "The ping was submitted after having been scheduled earlier than 4am today." or something that makes it about the value that's in this ping.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt (hunk 39 above)
Comment (mdboom) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/private/LabeledMetricTypeTest.kt]: Oops.  Fixed now.

File: glean-core/pings.yaml (hunk 41 above)
Comment (mdboom) [reply glean-core/pings.yaml]: Good point.  I've rephrased this in terms of "last ping" and "this ping", which hopefully makes things clearer.
